Zend/zend_errors.c

```diff
@@ -0,0 +1,203 @@
+				}
+			}
+		}
+	}
+#if ZEND_DEBUG
+	if (PG(report_zend_debug)) {
+		zend_bool trigger_break;
+
+		switch (type) {
+			case E_ERROR:
+			case E_CORE_ERROR:
+			case E_COMPILE_ERROR:
+			case E_USER_ERROR:
+				trigger_break=1;
+				break;
+			default:
+				trigger_break=0;
+				break;
+		}
+		zend_output_debug_string(trigger_break, "%s(%d) : %s - %s", error_filename, error_lineno, error_type_str, buffer);
+	}
+#endif
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ zend_error_log_cb */
+ZEND_ERROR_CB_API zend_error_log_cb(ZEND_ERROR_CB_HOOK_ARGS)
+{
+	char *buffer = PG(last_error_message);
+
+	if (!php_get_module_initialized() || PG(log_errors)) {
+		char *log_buffer;
+#ifdef PHP_WIN32
+		if (type == E_CORE_ERROR || type == E_CORE_WARNING) {
+			syslog(LOG_ALERT, "PHP %s: %s (%s)", error_type_str, buffer, GetCommandLine());
+		}
+#endif
+		spprintf(&log_buffer, 0, "PHP %s:  %s in %s on line %d", error_type_str, buffer, error_filename, error_lineno);
+		php_log_err(log_buffer);
+		efree(log_buffer);
+	}
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ zend_error_bailout_cb */
+ZEND_ERROR_CB_API zend_error_bailout_cb(ZEND_ERROR_CB_HOOK_ARGS)
+{
+	/* Bail out if we can't recover */
+	switch (type) {
+		case E_CORE_ERROR:
+			if(!php_get_module_initialized()) {
+				/* bad error in module startup - no way we can live with this */
+				exit(-2);
+			}
+		/* no break - intentionally */
+		case E_ERROR:
+		case E_RECOVERABLE_ERROR:
+		case E_PARSE:
+		case E_COMPILE_ERROR:
+		case E_USER_ERROR:
+			EG(exit_status) = 255;
+			if (php_get_module_initialized()) {
+				if (!PG(display_errors) &&
+				    !SG(headers_sent) &&
+					SG(sapi_headers).http_response_code == 200
+				) {
+					sapi_header_line ctr = {0};
+
+					ctr.line = "HTTP/1.0 500 Internal Server Error";
+					ctr.line_len = sizeof("HTTP/1.0 500 Internal Server Error") - 1;
+					sapi_header_op(SAPI_HEADER_REPLACE, &ctr);
+				}
+				/* the parser would return 1 (failure), we can bail out nicely */
+				if (type == E_PARSE) {
+					CG(parse_error) = 0;
+				} else {
+					/* restore memory limit */
+					zend_set_memory_limit(PG(memory_limit));
+					zend_objects_store_mark_destructed(&EG(objects_store));
+					zend_bailout();
+					return FAILURE;
+				}
+			}
+	}
+	return SUCCESS;
+}
+/* }}} */
+
+/* {{{ zend_append_error_hook */
+ZEND_API void zend_append_error_hook(zend_error_cb_hook_t hook_part, void (*hook)(ZEND_ERROR_CB_HOOK_ARGS))
+{
+	zend_llist_add_element(&PG(error_hooks)[hook_part], (void *)&hook);
+}
+/* }}} */
+
+/* {{{ zend_prepend_error_hook */
+ZEND_API void zend_prepend_error_hook(zend_error_cb_hook_t hook_part, void (*hook)(ZEND_ERROR_CB_HOOK_ARGS))
+{
+	zend_llist_prepend_element(&PG(error_hooks)[hook_part], (void *)&hook);
+}
+/* }}} */
+
+/* {{{ zend_clear_error_hook */
+ZEND_API void zend_clear_error_hook(zend_error_cb_hook_t hook_part)
```

Comment:
Cfr. mailing list discussion.

This is required when an extension is supposed to overtake the whole responsibility of a hook category by clearing all existing hooks in it and registering it's own one.
It is really more dangerous than the current situation that implies replacing the whole implementation of `php_error_cb`?